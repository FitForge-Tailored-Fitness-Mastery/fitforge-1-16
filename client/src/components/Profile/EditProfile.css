.edit-profile-container {
    max-width: 400px;
    min-width: 40vh;
    min-height: 100vh;
    margin: auto;
    padding: 20px;
    background-color: #333;
    color: white;
    font-family: Arial, sans-serif;
    border-radius: 10px;
  }
  
  .edit-profile-container h2 {
    text-align: center;
    margin-bottom: 20px;
  }
  
  .form-group {
    margin-bottom: 20px;
  }
  
  .form-group label {
    display: block;
    margin-bottom: 5px;
  }
  
  .form-group input {
    width: 70%;
    padding: 10px;
    border-radius: 5px;
    border: none;
  }
  
  button[type='submit'] {
    display: block;
    width: 100%;
    padding: 10px;
    background-color: gold;
    color: black;
    font-weight: bold;
    border: none;
    border-radius: 5px;
    cursor: pointer;
  }

  .edit-profile-container {
    max-width: 400px;
    margin: 0 auto;
    padding: 20px;
    background-color: #333;
    color: white;
    font-family: Arial, sans-serif;
    border-radius: 10px;
  }
  
  @media (max-width: 896px) {
    .profile-container {
      max-width: 90%;
      padding: 10px;
    }
  }
  
  @media (max-width: 480px) {
    .profile-container {
      margin: 20px 10px;
      padding: 10px;
    }
  }