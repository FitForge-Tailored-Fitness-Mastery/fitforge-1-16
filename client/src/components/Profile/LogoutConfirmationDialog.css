/* This is for the semi-transparent backdrop that covers the entire viewport */
.backdrop {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  padding: 20px; /* Add padding to ensure the dialog doesn't touch the viewport edges */
  background-color: rgba(0, 0, 0, 0.6);
  display: flex;
  justify-content: center;
  align-items: center;
  z-index: 1000; /* Ensure it's above other content */
}


/* This styles the actual dialog box */
.logout-dialog {
  max-width: 100px;
  background-color: #FFF; /* Use the color you want for the dialog */
  padding: 20px;
  border-radius: 10px;
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 60%; /* You can adjust the width as needed */
  max-width: 280px; /* Set a max-width for your dialog */
  box-shadow: 0 2px 10px rgba(0, 0, 0, 0.2); /* Optional: add some shadow for depth */
}

/* Styles for the text in the dialog */
.logout-dialog h2 {
  color: #333; /* Text color for the heading */
  margin-bottom: 16px; /* Space below the heading */
}

.logout-dialog p {
  color: #333; /* Text color for the paragraph */
  margin-bottom: 24px; /* Space below the paragraph */
}

/* Styles for the buttons and the container holding them */
.logout-dialog-buttons {
  display: flex;
  justify-content: center; /* Center buttons horizontally */
  width: 100%; /* Make the buttons container take full width of the dialog */
  gap: 16px; /* Space between buttons */
}

.logout-dialog-button {
  border: 1px solid #ccc; /* Button border */
  background-color: #fff; /* Button background color */
  padding: 10px 20px; /* Button padding */
  border-radius: 5px; /* Rounded corners for buttons */
  cursor: pointer; /* Cursor to pointer to indicate button */
  color: #333; /* Button text color */
  font-weight: bold; /* Make the font bold */
  flex: 1; /* Distribute space evenly */
}

/* Button hover effect */
.logout-dialog-button:hover {
  background-color: gold; /* Light background color on hover */
  border-color: #bbb; /* Darker border on hover */
}
