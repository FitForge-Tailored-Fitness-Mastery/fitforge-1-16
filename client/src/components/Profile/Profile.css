body{
    background-color: #1f1e1e;
    padding: 0;
    font-family: Georgia, 'Times New Roman', Times, serif;
}

/* Styles for the Navigation.js */
nav{
    background-color: #fdd673;
    color: white;
    display: flex;
    justify-content: space-around;
    align-items: center;
    padding: 0px;
    position:fixed;
    bottom: 0;
    width: 100%;
    border-radius: 5px;
}

.navbar-link{
    color: white;
    padding: 10px;
    text-decoration: none;
    font-size: 2.5rem;
    transition: color 0.3s;
}

.navbar-link:hover {
    color: black;
}

.navbar-link:active {
    color: black;
}


body {
  display: flex;
  justify-content: center; /* Centers horizontally */
  align-items: center; /* Centers vertically */
  min-height: 100vh;
  margin: 0; /* Removes the default margin */
}  

.profile-container {
  max-width: 400px;
  margin: auto; /* Applies auto margins on all sides, helping with centering */
  padding: 10px;
  background-color: #333;
  color: white;
  font-family: Arial, sans-serif;
  border-radius: 10px;
  display: flex;
  flex-direction: column;
  align-items: center; /* Ensures content inside .profile-container is centered */
  box-sizing: border-box;
}

.profile-header {
  text-align: center;
  margin-bottom: 20px;
}

.profile-picture-container {
  width: 150px;
  height: 150px;
  border-radius: 50%;
  overflow: hidden;
  margin-bottom: 20px;
}

.profile-photo {
  max-width: 100%;
  max-height: 100%;
  object-fit: cover;
}

.profile-name {
  color: gold;
}

.profile-details {
  display: flex;
  justify-content: space-around;
  margin-bottom: 20px;
  width: calc(100% - 40px);
  padding: 0 20px;
}

.detail-box {
  background-color: #555;
  padding: 20px;
  margin: 5px;
  border-radius: 10px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  flex: 1;
}

.detail-label, .detail-value {
  margin: 5px 0;
}

.edit-button {
  padding: 15px 20px;
  background-color: white;
  border: none;
  color: black;
  font-weight: bold; 
  width: 90%;
  margin: 0 auto 20px;
  display: block;
}

.menu-container {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
  gap: 10px; /* Space between "My Trainer" and "Logout" buttons */
}

.menu-button {
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 90%;
  padding: 15px 20px;
  background-color: #444;
  font-size: 1rem;
  color: white;
  cursor: pointer;
  border-radius: 10px;
  text-decoration: none;
  border: none;
  outline: none;
  margin-bottom: 10px;
  box-sizing: border-box;
}

.menu-arrow {
  margin-left: auto;
}

.error-message {
  color: red;
  font-weight: bold;
  margin-bottom: 10px;
}

@media (max-width: 896px) {
  .profile-container {
    max-width: 90%;
    padding: 10px;
  }
}

@media (max-width: 480px) {
  .profile-container {
    margin: 20px 10px;
    padding: 10px;
  }
}
