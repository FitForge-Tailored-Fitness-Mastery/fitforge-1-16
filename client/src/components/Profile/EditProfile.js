import React, { useState, useEffect } from 'react';
import { useParams, useNavigate } from 'react-router-dom';
import NavigationBar from '../NavigationBar/NavigationBar';
import './EditProfile.css';

const EditProfile = () => {
  const { clientId } = useParams();
  const navigate = useNavigate();
  const [client, setClient] = useState(null);
  const [formData, setFormData] = useState({
    fname: '',
    lname: '',
    height: '',
    dob: '',
    weight: '',
  });
  const [error, setError] = useState(null); // Adding state to handle errors

  useEffect(() => {
    const fetchClientData = async () => {
      try {
        const response = await fetch(`http://localhost:5000/client/${clientId}`);
        if (!response.ok) {
          throw new Error(`HTTP error! status: ${response.status}`);
        }
        const data = await response.json();
        setClient(data);
        setFormData({
          fname: data.fname,
          lname: data.lname,
          height: data.height,
          dob: data.dob,
          weight: data.weight,
        });
      } catch (error) {
        console.error('Fetch error:', error);
        setError('Failed to load data');
      }
    };
    fetchClientData();
  }, [clientId]);

  const handleInputChange = (e) => {
    let { name, value } = e.target;
    console.log(e, name, value)
    if (name === 'dob') {
      const date = new Date(value);
      value = date.toISOString().slice(0, 10); // Converts date to YYYY-MM-DD format
      console.log('Formatted Date:', value); // Check the formatted date
    }
    setFormData({ ...formData, [name]: value });
  };
    

  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      const response = await fetch(`http://localhost:5000/client/${clientId}`, {
        method: 'PATCH',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(formData),
      });
      if (!response.ok) {
        throw new Error(`HTTP error! Status: ${response.status}`);
      }
      const data = await response.json();
      console.log('Client updated:', data);
      navigate('/profile'); // Navigate back to the profile page after successful update
    } catch (error) {
      console.error('Update error:', error);
      setError(`Failed to update data: ${error.message}`); // Include the error message in the state
    }
  };  

  if (!client) {
    return <div>Loading...</div>;
  }

  return (
    <div className="edit-profile-container">
      <div className='edit-profile-box'>
          <h2>Edit Profile</h2>
          {/* Displaying error message */}
          {error && <p className="error">{error}</p>}
          <form id='editprofileform' onSubmit={handleSubmit}>
            <div className="form-group">
              <label htmlFor="fname">First Name:</label>
              <input
                type="text"
                id="fname"
                name="fname"
                value={formData.fname}
                onChange={handleInputChange}
              />
            </div>
            <div className="form-group">
              <label htmlFor="lname">Last Name:</label>
              <input
                type="text"
                id="lname"
                name="lname"
                value={formData.lname}
                onChange={handleInputChange}
              />
            </div>
            <div className="form-group">
              <label htmlFor="height">Height (cm):</label>
              <input
                type="number"
                id="height"
                name="height"
                value={formData.height}
                onChange={handleInputChange}
              />
            </div>
            <div className="form-group">
              <label htmlFor="dob">Date of Birth:</label>
              <input
                type="date"
                id="dob"
                name="dob"
                value={formData.dob}
                onChange={handleInputChange}
              />
            </div>
            <div className="form-group">
              <label htmlFor="weight">Weight (lb):</label>
              <input
                type="number"
                id="weight"
                name="weight"
                value={formData.weight}
                onChange={handleInputChange}
              />
            </div>
            <button type="submit" className='editprofilebutton'>Save Changes</button>
          </form>
      </div>
      {/* Add Navigation Bar */}
      <NavigationBar />
    </div>
  );
};

export default EditProfile;
