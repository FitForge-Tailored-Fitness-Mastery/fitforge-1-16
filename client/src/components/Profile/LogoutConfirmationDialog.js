import React from 'react';
import { useNavigate } from 'react-router-dom'; // Import useNavigate
import './LogoutConfirmationDialog.css';

const LogoutConfirmationDialog = ({ onClose, onConfirm }) => {
  const navigate = useNavigate(); // Create the navigate function

  // Update the onConfirm function to navigate to login
  const handleConfirm = () => {
    onConfirm(); // Call the original onConfirm function if needed
    navigate('/login'); // Navigate to the login page
  };

  return (
    <div className="backdrop">
      <div className="logout-dialog">
        <h2>Going so soon?</h2>
        <p>Are you sure you want to Logout?</p>
        <div className="logout-dialog-buttons">
          <button className="logout-dialog-button" onClick={onClose}>
            Cancel
          </button>
          <button className="logout-dialog-button" onClick={handleConfirm}>
            Yes
          </button>
        </div>
      </div>
    </div>
  );
};

export default LogoutConfirmationDialog;
