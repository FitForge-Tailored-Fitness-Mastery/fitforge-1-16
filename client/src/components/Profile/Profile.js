import React, { useState, useEffect } from 'react';
import './Profile.css';
import profileImage from './proimg.jpg';
import LogoutConfirmationDialog from './LogoutConfirmationDialog';
import NavigationBar from '../NavigationBar/NavigationBar';
import '../ClientHome/ClientHome.css'; // Import associated styles for navbar
import { Link, useNavigate } from 'react-router-dom';

const Profile = ({ clientId }) => {
  const [showLogoutConfirmation, setShowLogoutConfirmation] = useState(false);
  const [clientData, setClientData] = useState(null);
  const [error, setError] = useState(null);
  const navigate = useNavigate();

  const fetchClientData = async (client_id) => {
    try {
      const response = await fetch(`http://localhost:5000/client/${client_id}`);
      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }
      const data = await response.json();
      console.log(data);
      setClientData(data);
    } catch (error) {
      console.error('Fetch error:', error);
      setError('Failed to load client data');
    }
  };

  useEffect(() => {
    const clientId = localStorage.getItem('clientId');
    fetchClientData(String(clientId));
  }, []);

  const handleLogoutClick = () => {
    setShowLogoutConfirmation(true);
  };

  const handleCancelLogout = () => {
    setShowLogoutConfirmation(false);
  };

  const handleConfirmLogout = () => {
    setShowLogoutConfirmation(false);
    localStorage.removeItem('clientId');
  };

  const handleEditProfile = () => {
    navigate(`/edit-profile/${clientData.client_id}`);
  };

  function calculateAge(dob) {
    const birthday = new Date(dob);
    const ageDifMs = Date.now() - birthday.getTime();
    const ageDate = new Date(ageDifMs);
    return Math.abs(ageDate.getUTCFullYear() - 1970);
  }

  return (
    <div className="profile-container">
      {/* Render error message if there is an error */}
      {error && <div className="error-message">{error}</div>}
      {/* Render profile details only if there is no error and clientData is available */}
      {!error && clientData && (
        <>
          <div className="profile-header">
            <div className="profile-picture-container">
              <img src={profileImage} alt="User" className="profile-photo" />
            </div>
            <h2 className="profile-name">
              {clientData.fname} {clientData.lname}
            </h2>
          </div>
          <div className="profile-details">
            <div className="detail-box">
              <p className="detail-label">Height</p>
              <p className="detail-value">{clientData.height} cm</p>
            </div>
            <div className="detail-box">
              <p className="detail-label">Age</p>
              {/* Make sure you calculate the age from dob */}
              <p className="detail-value">{calculateAge(clientData.dob)}</p>
            </div>
            <div className="detail-box">
              <p className="detail-label">Weight</p>
              <p className="detail-value">{clientData.weight} lb</p>
            </div>
          </div>
          <button className="edit-button" onClick={handleEditProfile}>
            Edit Profile
          </button>
          <div className="menu-container">
            <Link
              to={`/trainerprofile/${clientData.trainer_id}`}
              className="menu-button"
              style={{ textDecoration: 'none' }}
            >
              My Trainer <span className="menu-arrow">{' >'}</span>
            </Link>
            <button className="menu-button" onClick={handleLogoutClick}>
              Logout <span className="menu-arrow">{' >'}</span>
            </button>
          </div>
        </>
      )}
      {showLogoutConfirmation && (
        <LogoutConfirmationDialog
          onClose={handleCancelLogout}
          onConfirm={handleConfirmLogout}
        />
      )}
      
      {/* Add Navigation Bar */}
      <NavigationBar />
      
    </div>
  );
};

export default Profile;