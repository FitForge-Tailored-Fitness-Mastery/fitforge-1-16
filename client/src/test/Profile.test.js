import React from 'react';
import { render, screen, waitFor, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import Profile from '../components/Profile/Profile';
import { BrowserRouter } from 'react-router-dom'; // Add this import
import renderer from 'react-test-renderer';

// Mock fetch
global.fetch = jest.fn();

describe('Profile', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('displays loading state initially', async () => {
    const tree = renderer.create(<BrowserRouter><Profile /></BrowserRouter>).toJSON();
    Â  expect(tree).toMatchSnapshot();
  });

  it('displays the client details when fetching is successful', async () => {
    const mockClientData = {
      fname: 'John',
      lname: 'Doe',
      height: 180,
      weight: 160,
      dob: '1990-01-01',
      trainer_id: '456',
    };

    global.fetch.mockResolvedValueOnce({
      ok: true,
      json: async () => mockClientData,
    });

    render(
      <BrowserRouter> {/* Wrap the component with BrowserRouter */}
        <Profile />
      </BrowserRouter>
    );

    await waitFor(() => {
      expect(screen.getByText('John Doe')).toBeInTheDocument();
      expect(screen.getByText('180 cm')).toBeInTheDocument();
      expect(screen.getByText('34')).toBeInTheDocument(); // Assuming current year is 2024
      expect(screen.getByText('160 lb')).toBeInTheDocument();
    });
  });

  it('displays an error message when the fetch fails', async () => {
    global.fetch.mockRejectedValueOnce(new Error('Failed to fetch'));

    render(
      <BrowserRouter> {/* Wrap the component with BrowserRouter */}
        <Profile />
      </BrowserRouter>
    );

    await waitFor(() => {
      expect(screen.getByText('Failed to load client data')).toBeInTheDocument();
    });
  });
});
