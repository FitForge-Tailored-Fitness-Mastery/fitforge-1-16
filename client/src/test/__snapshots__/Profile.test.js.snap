// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile displays loading state initially 1`] = `
<div
  className="profile-container"
>
  <nav
    className="navbar"
    data-testid="navigation"
  >
    <a
      className="navbar-link"
      href="/workout-history"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-dumbbell "
        data-icon="dumbbell"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 640 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 64c0-17.7 14.3-32 32-32h32c17.7 0 32 14.3 32 32V224v64V448c0 17.7-14.3 32-32 32H128c-17.7 0-32-14.3-32-32V384H64c-17.7 0-32-14.3-32-32V288c-17.7 0-32-14.3-32-32s14.3-32 32-32V160c0-17.7 14.3-32 32-32H96V64zm448 0v64h32c17.7 0 32 14.3 32 32v64c17.7 0 32 14.3 32 32s-14.3 32-32 32v64c0 17.7-14.3 32-32 32H544v64c0 17.7-14.3 32-32 32H480c-17.7 0-32-14.3-32-32V288 224 64c0-17.7 14.3-32 32-32h32c17.7 0 32 14.3 32 32zM416 224v64H224V224H416z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </a>
    <a
      className="navbar-link active"
      href="/client-home"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-house "
        data-icon="house"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 576 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M575.8 255.5c0 18-15 32.1-32 32.1h-32l.7 160.2c0 2.7-.2 5.4-.5 8.1V472c0 22.1-17.9 40-40 40H456c-1.1 0-2.2 0-3.3-.1c-1.4 .1-2.8 .1-4.2 .1H416 392c-22.1 0-40-17.9-40-40V448 384c0-17.7-14.3-32-32-32H256c-17.7 0-32 14.3-32 32v64 24c0 22.1-17.9 40-40 40H160 128.1c-1.5 0-3-.1-4.5-.2c-1.2 .1-2.4 .2-3.6 .2H104c-22.1 0-40-17.9-40-40V360c0-.9 0-1.9 .1-2.8V287.6H32c-18 0-32-14-32-32.1c0-9 3-17 10-24L266.4 8c7-7 15-8 22-8s15 2 21 7L564.8 231.5c8 7 12 15 11 24z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </a>
    <a
      className="navbar-link"
      href="/profile"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-user "
        data-icon="user"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512H418.3c16.4 0 29.7-13.3 29.7-29.7C448 383.8 368.2 304 269.7 304H178.3z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </a>
  </nav>
</div>
`;
